readme.md
# **wzrySpider**

## **Basic Requirements**

### ***Python versions***

应该安装好Python(v3.4.3或者更高，可以使用Python -V来确认)。

### ***Install PhantomJS***

PhantomJS安装方法有两种，一种是下载源码之后自己来编译，另一种是直接下载编译好的二进制文件。然而自己编译需要的时间太长，而且需要挺多的磁盘空间。官方推荐直接下载二进制文件然后安装。

大家可以依照自己的开发平台选择不同的包进行下载

[下载地址](http://phantomjs.org/download.html)

当然如果你不嫌麻烦，可以选择

[下载源码](http://phantomjs.org/build.html)

### ***Install webdriver(可选)***

PhantomJS是一个无头浏览器，没有UI界面，所以建议再下一个火狐驱动或者谷歌驱动，这样你可以看到页面的跳转等具体是一个什么样子。

- Firefox的驱动geckodriver 下载地址：

    <https://github.com/mozilla/geckodriver/releases/>

- chromium的驱动chromedriver 下载地址:

    <https://sites.google.com/a/chromium.org/chromedriver/downloads>

### ***Using Python module:***

***方法一：***

- [selenium](www.seleniumhq.org)

        $ pip install selenium

- [BeautifulSoup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)

        $ pip install beautifulsoup4

- [requests](www.python-requests.org)

        $ pip install requests

***方法二：***

如果你已经安装好pipenv,那么可以直接在项目下面执行(可选：

    $ pipenv install
    $ pipenv shell

***方法三：***

项目下有requirements.txt文件，因此可以直接：

    $ pip install -r requirements.txt

## **Run program**

    $ python3 main.py

## **BUGS**

- 有四张照片会被爬取很多遍，这是网页决定的，可以选择链接去重，因为相同名字的图片会被覆盖，所以这里就没有去管

CHANGELOG.md
# **History**

## **v1.0.0 (2018-1-16)**

- init

## **v1.0.1 (2018-1-31)**

- support pipenv and yapf

## **v1.0.2 (2018-1-31)**

- add requirements.txt, Modify readme.md.


test/wzry-wall-dl.py
########################################################
#        Program to Download Wallpapers from           #
#     http://pvp.qq.com/web201605/wallpaper.shtml      #
#                                                      #
#                 Author - CQ                          #
#                                                      #
#                 dated - 2018-1-8 14:28:52            #
#                 Update - 2018-1-8 14:28:57           #
########################################################

import os
import time
from selenium import webdriver
from selenium.webdriver.common.action_chains import ActionChains
from bs4 import BeautifulSoup
import requests

driver = webdriver.PhantomJS()
driver.maximize_window()

url = 'http://pvp.qq.com/'
driver.get(url)
moveElement = driver.find_element_by_xpath('//a[@title="游戏资料"]')
ActionChains(driver).move_to_element(moveElement).perform()
driver.find_element_by_xpath('//a[@title="游戏壁纸"]').click()
all_h = driver.window_handles
print(all_h)
driver.switch_to.window(all_h[1])
time.sleep(3)

if os.path.isdir("D://王者荣耀"):
    os.chdir("D://王者荣耀")
else:
    os.mkdir("D://王者荣耀")
    os.chdir("D://王者荣耀")

page_num = 0

while page_num < 12:
    soup = BeautifulSoup(driver.page_source, 'html.parser')
    Lists = soup.find_all('div', {'class': 'p_newhero_item'})
    titleLists = []
    hrefLists = []
    for item in Lists:
        subSoup = BeautifulSoup(str(item), 'html.parser')
        titleList = subSoup.find('h4')
        titleLists.append(titleList.text)
        linkList = subSoup.find('li', {'class': 'sProdImgL6'})
        soup = BeautifulSoup(str(linkList), 'html.parser')
        a = soup.find('a')
        hrefLists.append(a['href'])

    for i in range(len(titleLists)):
        url = hrefLists[i]
        r = requests.get(url).content
        path = titleLists[i] + '.jpg'
        try:
            with open(path, 'wb') as f:
                f.write(r)
                print('保存成功 %s.jpg' % titleLists[i])
        except BaseException:
            print('保存失败')

    driver.find_element_by_xpath('//a[@class="downpage"]').click()
    time.sleep(3)
    page_num = page_num + 1
    print(page_num)
driver.close()


main.py
"""
########################################################
#        Program to Download Wallpapers from           #
#     http://pvp.qq.com/web201605/wallpaper.shtml      #
#                                                      #
#    reference: https://www.jianshu.com/p/395354e3d9ef #
#                                                      #
#                 Author - CQ                          #
#                                                      #
#                 dated - 2018-1-8 14:28:52            #
#                 Update - 2018-1-16 14:54:30          #
########################################################

"""
from src.app.wzry_wall_dl import WzrzWallDL

def main():
    """mail"""
    wzrz_wall_dl = WzrzWallDL()
    wzrz_wall_dl.wzrz_wall_dl()

if __name__ == '__main__':
    main()


src/comm/utils.py
# -*- coding: utf-8 -*-

"""
created by：2017-05-10 20:11:31
modify by: 2018-1-16 10:39:45

功能：各种常用的方法函数的封装。
"""

import os
import json


class Utils:
    """util, 工具类

    Attributes:

    """

    @staticmethod
    def rename_file(oldfile, newfile):
        """文件重命名

        参数:

            oldfile：旧文件名；string。
            newfile：新文件名;string。
        """
        if not os.path.isfile(newfile):
            os.rename(oldfile, newfile)

    @classmethod
    def get_path_of_file(cls, dir_path):
        """遍历文件夹, 返回所有文件list(不含有文件夹)"""
        if os.path.exists(dir_path):
            file_path_list = []
            for root, dirs, files in os.walk(dir_path):
                file_path_list.extend(os.path.join(root, file) for file in files)
            return file_path_list
        else:
            print("This path %s does not exist!" % (dir_path))

    @classmethod
    def get_listdir(cls, dir_path):
        """遍历文件夹,单层"""
        if os.path.exists(dir_path):
            value_list = os.listdir(dir_path)
            return value_list
        else:
            print("This path %s does not exist!" % (dir_path))

    @staticmethod
    def load_json_file(json_file):
        """加载json文件"""
        with open(json_file, "r", encoding="utf8")  as frs:
            res = json.load(frs)
        return res

    @classmethod
    def dump_json_file(cls, data, json_file, ensure_ascii=False):
        """写入数据json文件"""
        if data is not None:
            res = json.dumps(data, ensure_ascii=ensure_ascii)
            with open(json_file, "w", encoding="utf8")  as fws:
                fws.write(res)
        else:
            print("This json data is None")

    @staticmethod
    def get_json_value(json_file, key):
        """根据key获取json文件中的value"""
        with open(json_file, "r", encoding="utf8")  as frs:
            res = json.load(frs)
        if key in res:
            return res[key]

    @staticmethod
    def set_json_value(json_file, key, value):
        """根据key设置json文件中的value"""
        with open(json_file, "r", encoding="utf8")  as frs:
            res = json.load(frs)
        res[key] = value
        result = json.dumps(res, indent=4, sort_keys=True, ensure_ascii=False).replace("'", "\"")
        with open(json_file, "r", encoding="utf8")  as fws:
            fws.write(result)


src/comm/logging_util.py
# -*- coding: utf-8 -*-

"""
created by：2017-01-10 20:11:31
modify by: 2018-1-16 14:49:22

功能：logging模块常用两种日志轮询方法的封装。
"""

import logging
from logging import StreamHandler
from logging.handlers import RotatingFileHandler, TimedRotatingFileHandler

class SingletonPattern(type):
    """Singleton Pattern"""
    _inst = {}

    def __call__(cls, *args, **kwargs):
        if cls not in cls._inst:
            cls._inst[cls] = super(SingletonPattern, cls).__call__(*args)
        return cls._inst[cls]

class LoggingFileHandler(SingletonPattern):
    """logging 日志轮询二次封装，工具类 """
    @staticmethod
    def size_rotating_file_handler(logfile, username, log_format=None, log_encoding=None,
                                   log_datefmt=None, max_bytes=10*1024*1024,
                                   level=logging.INFO, backup_count=5):
        """根据文件大小切割日志"""

        # 设置日志格式
        if log_format is None:
            log_format = '%(asctime)-8s - %(name)5s: %(filename)5s  %(levelname)-8s %(message)s'

        # 设置时间格式
        if log_datefmt is None:
            log_datefmt = '%Y-%m-%d %H:%M:%S'

        # 生成一个日志对象
        logger = logging.getLogger(username)
        # 设置日志级别
        logger.setLevel(level)
        log_formatter = logging.Formatter(fmt=log_format, datefmt=log_datefmt)

        rthandler = RotatingFileHandler(logfile, maxBytes=max_bytes,
                                        backupCount=backup_count, encoding=log_encoding)
        rthandler.setFormatter(log_formatter)
        logger.addHandler(rthandler)

        console_handler = StreamHandler()
        console_handler.setFormatter(log_formatter)
        logger.addHandler(console_handler)
        return logger

    @staticmethod
    def timed_rotating_file_handler(logfile, username, log_format=None, log_encoding=None,
                                    log_datefmt=None, rotating_time="midnight",
                                    level=logging.INFO, backup_count=5,):
        """根据时间切割日志"""

        # 设置日志格式
        if log_format is None:
            log_format = '%(asctime)-8s - %(name)5s: %(filename)5s  %(levelname)-8s %(message)s'
        # 设置时间格式
        if log_datefmt is None:
            log_datefmt = '%Y-%m-%d %H:%M:%S'

        # 生成一个日志对象
        logger = logging.getLogger(username)
        # 设置日志级别
        logger.setLevel(level)
        log_formatter = logging.Formatter(fmt=log_format, datefmt=log_datefmt)

        trthandler = TimedRotatingFileHandler(logfile, when=rotating_time,
                                              backupCount=backup_count, encoding=log_encoding)
        trthandler.setFormatter(log_formatter)
        logger.addHandler(trthandler)

        console_handler = StreamHandler()
        console_handler.setFormatter(log_formatter)
        logger.addHandler(console_handler)
        return logger

if __name__ == '__main__':
    pass


src/app/wzry_wall_dl.py
# -*- coding: utf-8 -*-

"""
created by：2018-1-14 11:38:2
modify by: 2018-1-31 18:27:47

功能：王者荣耀壁纸下载
"""

import sys
import os
import time

import requests
from selenium import webdriver
from selenium.webdriver.common.action_chains import ActionChains
from bs4 import BeautifulSoup

from src.comm.app_config import *
from src.comm.utils import Utils
from src.comm.logging_util import LoggingFileHandler

class WzrzWallDL:
    """
    王者荣耀壁纸下载
    """

    def __init__(self):
        self.conf_path = os.path.join(os.getcwd(), "res/conf/config.json")

        if os.path.isfile(self.conf_path):
            conf_json_info = Utils.load_json_file(self.conf_path)
            print("OK: The config file is %s !!!" % (self.conf_path))
            self.url = conf_json_info["config"]["url"]
            self.max_page_number = conf_json_info["config"]["max_page_number"]
            self.download_path = conf_json_info["config"]["download_path"]
            self.images_resolution = conf_json_info["config"]["images_resolution"]
            logfile_path = conf_json_info["config"]["logfile_path"]
            program_name = conf_json_info["config"]["program_name"]
        else:
            print("Unkown: The config file %s is not exit!!!" % (self.conf_path))
            sys.exit(1)

        self.logger_handle = LoggingFileHandler.timed_rotating_file_handler(logfile_path,
                            program_name, log_encoding="utf-8")

    def wzrz_wall_dl(self):
        # 获得浏览器的session，浏览器用Firefox、Chrome、IE等都可以
        # driver = webdriver.Chrome()
        driver = webdriver.PhantomJS()
        # 浏览器全屏显示
        driver.maximize_window()
        # 加载页面
        driver.get(self.url)
        """
        # 获得了session对象后，要定位元素，webdriver提供了一系列的元素定位方法，常用的有以下几种方式:
        # id, name, class-name, link, text, partial, link, text, tag, name, xpath, cssselector
        """
        move_element = driver.find_element_by_xpath('//a[@title="游戏资料"]')

        """
        # 当你调用ActionChains的方法时，不会立即执行，而是会将所有的操作按顺序存放在一个队列里，当你调用perform()方法时，队列中的时间会依次执行。
        # move_to_element(to_element) —— 鼠标移动到某个元素
        # click(on_element=None) —— 单击鼠标左键
        """
        ActionChains(driver).move_to_element(move_element).perform()

        driver.find_element_by_xpath('//a[@title="游戏壁纸"]').click()
        # 获取所有的窗口
        all_handles = driver.window_handles
        # 跳入窗口
        driver.switch_to.window(all_handles[1])
        time.sleep(3)

        if os.path.isdir(self.download_path):
            os.chdir(self.download_path)
        else:
            os.mkdir(self.download_path)
            os.chdir(self.download_path)

        # 初始页数
        page_num = 0

        if self.images_resolution in images_resolution_list:
            dl_images_resolution = images_resolution_list[self.images_resolution]
        else:
            self.logger_handle.error("The images_size: %s is not exits!!!", self.images_resolution)
            driver.close()
            sys.exit(1)

        while page_num <= self.max_page_number:
            soup = BeautifulSoup(driver.page_source, 'html.parser')
            newhero_item_lists = soup.find_all('div', {'class': 'p_newhero_item'})

            title_lists = []
            href_lists = []
            span_size_list = []

            for item in newhero_item_lists:
                sub_soup = BeautifulSoup(str(item), 'html.parser')
                title_list = sub_soup.find('h4')
                title_lists.append(title_list.text)
                link_list = sub_soup.find('li', {'class': dl_images_resolution})
                soup = BeautifulSoup(str(link_list), 'html.parser')
                a = soup.find('a')
                href_lists.append(a['href'])
                span_size_list.append(soup.span.string)

            for (i, x) in enumerate(title_lists):
                url = href_lists[i]
                rep = requests.get(url).content
                path = title_lists[i] + "_" + span_size_list[i] + '.jpg'
                try:
                    with open(path, 'wb') as imgage_file:
                        imgage_file.write(rep)
                except IOError as err:
                    self.logger_handle.error('保存失败 : %s', err)
                    self.logger_handle.error("The download url: %s , The filename: %s ", url, path)
                else:
                    self.logger_handle.info('保存成功 %s.jpg', title_lists[i])
                    self.logger_handle.info("The download url: %s , The filename: %s ", url, path)

            driver.find_element_by_xpath('//a[@class="downpage"]').click()
            time.sleep(3)
            page_num = page_num + 1

        # 当执行完抓取操作后，必须关闭session，不然让它一直占内存会影响机器其他进程的运行
        driver.close()
        print("OK: 运行完成，程序已经关闭!!!")
